[quant][graphmode] Prepack folding for conv2d

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27403 [quant][graphmode] Minor fix and turn off fold_convbn
* #27348 [quant][graphmode] Add e2e test for conv+bn
* #27347 [quant][graphmode][fix] Don't permute in quantized::conv2d pattern
* #27346 [quant][graphmode][fix] insert_prepack_unpack for conv
* #27269 [refactor] Move tests in test_quantizer.py to test_quantization.py
* #27239 [refactor] Factor out insertPrepackUnpackForLinear
* #27238 [refactor] InsertObservers
* #27237 [quant][graphmode][fix] Make inserted child module names unique
* **#27119 [quant][graphmode] Prepack folding for conv2d**

Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

jerryzh168 committed:
[WIP][quant][graphmode] Prepack folding for conv2d

Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -967,12 +994,27 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
       set_weight_bias(std::vector<IValue>{IValue(w_quant), IValue(b)});
       // TODO: we need to make sure this name is unique
-      module.register_module("_packed_linear_weight_bias", wrapper);
+      module.register_module("_packed_weight_bias", wrapper_module);
```
This will not work if a module has both `linear` and `conv2d` (or several `conv`s).

jerryzh168 replied:
yeah see the comment above, I'm planning to do this in a separate PR

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -967,12 +994,27 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
+          TORCH_WARN("Non-constant stride/padding/dilation");
+          continue;
+        }
+        set_conv_params(std::vector<IValue>{stride.value(), padding.value(), dilation.value(), groups.value()});
```
Nit: clang-format.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -938,19 +954,30 @@ void InsertPrepackUnpack(script::Module& module) {
     const std::string& method_name,
-    const script::Module& wrapper_module) {
+    const script::Module& linear_params_module,
+    const script::Module& conv_params_module) {
```
This looks a bit ugly. Are `linear` and `conv` the only ops that might need packing? Should we consider making this more general?

jerryzh168 replied:
these two should be the only cases, and these are workaround, I think we'll have proper support before we have more use cases.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -981,19 +1023,19 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
-      std::vector<Node*> nodes_to_delete = {w_val->node(),
-                                            b_val->node(),
+      std::vector<Node*> nodes_to_delete = {//w_val->node(),
+                                            //b_val->node(),
```
Why was this changed?

jerryzh168 replied:
I changed `%weight` and `%bias` to be the input of the pattern, when I try to delete them they have more uses because they are input of the method, so the uses is not empty for them and I can not delete them:
```
graph(%self : ClassType<Conv2d>,
      %input.1 : Tensor,
      %weight.1 : Tensor):
  %27 : float = prim::Constant[value=0.000733255]()
  %28 : int = prim::Constant[value=0]()
  %29 : int = prim::Constant[value=12]()
  %22 : float = prim::Constant[value=0.0783038]()
  %23 : int = prim::Constant[value=65]()
  %24 : int = prim::Constant[value=13]()
  %input.1.quant : Tensor = aten::quantize_per_tensor(%input.1, %22, %23, %24)
  %input.1.dequant : Tensor = aten::dequantize(%input.1.quant)
  %7 : None = prim::Constant() # /home/jerryzh/anaconda2/envs/ve3/lib/python3.6/site-packages/torch/nn/modules/conv.py:339:36
  %8 : int = prim::Constant[value=1]() # /home/jerryzh/anaconda2/envs/ve3/lib/python3.6/site-packages/torch/nn/modules/conv.py:336:46
  %9 : int[] = prim::Constant[value=[2, 2]]()
  %10 : int[] = prim::Constant[value=[3, 3]]()
  %11 : int[] = prim::Constant[value=[1, 1]]()
  %17 : float = prim::Constant[value=0.0275517]()
  %18 : int = prim::Constant[value=65]()
  %36 : ClassType<ConvPackedParams> = prim::GetAttr[name="_packed_weight_bias"](%self)
  %37 : Tensor = prim::GetAttr[name="_packed_params"](%36)
  %32 : Tensor, %33 : Tensor? = quantized::conv_unpack(%37)
  %34 : Tensor = aten::dequantize(%32)
  %35 : Tensor = aten::conv2d(%input.1.dequant, %34, %7, %9, %10, %11, %8)
  %19 : int = prim::Constant[value=13]()
  %12.quant : Tensor = aten::quantize_per_tensor(%35, %17, %18, %19)
  %12.dequant : Tensor = aten::dequantize(%12.quant)
  return (%12.dequant)
```

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -967,34 +997,54 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
-      auto packed_params_val = match_vmap.at(vmap.at("packed_params"));
+      // unique name for the module based on %w_quant
+      auto module_name =
+          module_name_prefix + std::to_string(w_quant_val->unique());
```
I suggest not depending on `unique()` and rather generate a unique name ourselves, but it's fine if you do it in a follow-up.

jerryzh168 replied:
this is fixed in follow up PR

ZolotukhinM reviewed (changes requested):
Mostly looks good, please find some remarks inline:

jerryzh168 commented on test/test_jit.py:
```diff
@@ -1410,30 +1406,50 @@ def forward(self, x):
+                              ('conv', QConv, torch.randn((1, 3, 24, 24), dtype=torch.float))]:
+            m = torch.jit.script(M())
+            ref_res = get_forward(m._c)(data)
+            fold_prepack(m)
```
yeah, you mean split the check in 1441 into a different file? but the behavior we want is these weights are packed properly in this context. I think there is still value to put this here

ZolotukhinM replied:
I meant testing the functionality of these classes directly without using our pass (e.g. using the class directly in Python it to pack, then unpack and verifying that the values are the same, maybe also packing and making sure the packed values are accepted by quantized kernels).

jerryzh168 replied:
I think this is more about checking the result of fold_prepack rather than test the functionality of the PackedParams modules. basically we want to make sure we insert the module and called `set_weight_bias` on correct values.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -638,6 +639,22 @@ graph(%a_dequant, %w, %b, %w_scale, %w_zero_point, %w_dtype, %stride, %padding,
   rewriter.runOnGraph(graph);
 }
 
+c10::optional<IValue> toIntList(Value* v) {
```
it does return intlist, do you mean we should add a check in L645 to check it's an IntList?

ZolotukhinM replied:
I think the new version is good.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -638,6 +639,22 @@ graph(%a_dequant, %w, %b, %w_scale, %w_zero_point, %w_dtype, %stride, %padding,
+    auto e1 = toIValue(n->inputs()[1]);
+    if (!e0 || !e1) {
+      return c10::nullopt;
+    }
```
because this is conv2d, I can add a check

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Prepack folding for conv2d"


Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

[ghstack-poisoned]
